### DISCUSSION TITLE
DSMR Fluvius S1 port reader

### DISCUSSION THREAD

**Discussion by BartVFutech** [tribp/DSMR-Fluvius-S1-port-simulator#4] [General]
Dag Peter,

Ik zie dat u reeds begonnen bent met het implementeren van een S1 poort lezer. Hoe ver staat u hiermee? En bent u bereid deze repository te delen? Indien gewenst wil ik ook helpen dit verder te implementeren.

Groeten,
Bart

**Comment by tribp** [comment]
Beste Bart,

Ik wil, als open community supporter :-), best mijn code delen.

Reden waarom dit (nog?) niet publiek stond was dat:
- code nog in experimenteer stadium was en nog niet opgeschoond
- ik uiteindelijk een beetje vastgelopen was bij gebrek aan een concrete toepassing.

PS: huidige code leest een S1 telegram en berekent Fast Fourier

Ik weet dat men theoretisch gezien uit de frequentie analyse van een 50hz signaal, op basis de Fourier ‘signatuur"  zou kunnen afleiden welk toestel er aangesloten is. Of meer algemeen de Power Quality zou kunnen berekenen maar tot op heden nog niemand gesproken die dit wat kon concreter maken.
 
Het afleiden welke toestel er aangesloten is kan ik vandaag relatief goed waarnemen via de P1 poort en de “signatuur” van het afgenomen vermogen.

Ik ben wel geen energie expert en ben wel benieuwd naar betere inzichten :-)

PS: momenteel ben ik wel actief met een open community initiatief om de P1 poort uit te lezen + solar productie + solar voorspelling zodoende bepaalde gebruikers “slim” te kunnen sturen

Mocht je nog extra vragen hebben of iets meer context hebben mag je me steeds bellen, of via teams

Mvg Peter

**Reply by BartVFutech** [reply]
Dag Peter,

Ik zou via de S1 poort graag de netfrequentie bekomen. Deze wil ik dan gebruiken om energieopslag aan te sturen ter ondersteuning van het net. Het uitlezen van de verschillende parameters volstaat dus voor mij. In de toekomst kan het wel zijn dat ik mij verder wil focussen om de data te analyse en toestellen te detecteren. Ik denk dus dat jouw code mij zeker verder kan helpen. 

Hoe dacht je de code te delen? Momenteel heb ik nog niet de rechten om de repository te openen.

Mvg Bart

**Comment by tribp** [comment]
Beste Bart,

Hierbij de code. Sorry dat alles nog niet opgeschoond en hopelijk help dit je een stuk op weg..

Succes.

Mvg Peter

**Reply by BartVFutech** [reply]
Dag Peter,

Ik heb het vermoeden dat je bijlage niet doorkomt via GitHub. Zou je het eens rechtstreeks naar mijn mailadres (bart.venmans@futech.be) kunnen proberen? 

Alvast bedankt!
Bart

**Comment by PerformerH** [comment]
Beste Peter,
Ik zou de data willen gebruiken om het vermogen beschikbaar te stellen via modbus. 1 meting per seconde gaat te traag. De s1 poort is hierbij een stuk performanter. kun je uw repository met me delen? 
Mvg, Hansie.

**Reply by tribp** [reply]
Beste Hansie,
Geen probleem, kan je me een e-mail bezorgen dan reply ik aangezien via Github geen attachments meegegeven kunnen worden
gr Peter

**Reply by PerformerH** [reply]
Hansiegoeman@hotmail.com 
Alvast bedankt.

**Comment by tribp** [comment]
Beste Hansie,

Sorry voor het late antwoord.

k wil, als open community supporter :-), best mijn code delen.

Reden waarom dit (nog?) niet publiek stond was dat:
- code nog in experimenteer stadium was en nog niet opgeschoond
- ik uiteindelijk een beetje vastgelopen was bij gebrek aan een concrete toepassing.

PS: huidige code leest een S1 telegram en berekent Fast Fourier

Veel succes

Mvg Peter

**Comment by divda** [comment]
Peter,

Ik ben ook geïnterresserd in je S1 poort lezer code. Ik zou die willen gebruiken om er een pvrouter ofte pv diverter mee te maken zoals o.a. bescherven in https://mk2pvrouter.co.uk/ en  https://learn.openenergymonitor.org/pv-diversion/mk2/index .
Het is mijn gedoeling om het gebruikte analoge meetcircuit voor spanning en stroom te vervangen door gegevens van de S1 interface. De P1 poort is veel te traag voor die toepassing.

Mvg, Dirk

**Comment by tribp** [comment]
Beste Dirk,

Code voor S1 receiver heb ik  toegevoegd in de "src" folder.
ps: code werkt maar is wel tesl/draft code :-)  en geen voorbeeld van "Pythonic" C code.

mvg Peter

**Reply by divda** [reply]
Bedankt voor de receiver code, en uiteraard ook voor de S1 poort simulator.

**Comment by PerformerH** [comment]
Beste Dirk,

Het uitlezen van de S1 poort werkt bij mij niet met arduino, ik heb hier heel wat testen mee gedaan maar weet niet waar het precies fout loopt. Ik krijg de RX buffer niet snel genoeg gelezen.

Waarschijnlijk staat er een instelling niet correct.
Met esp idf is het wel gelukt. Het is ook mijn bedoeling om er nog een dimmer voor waterverwarming mee te maken.

Bij mijn meter krijg ik 78 samples per Hz.

@Peter, lukt het jou om een volledige sinus in 1 keer uit te lezen met arduino?  Ik geef hier de voorkeur aan omdat ik er meer vertrouwd mee ben.

Wil je esp idf gebruiken,  dan kan ik mijn code wel even opschonen en doorsturen.

Mvg Hansie.




Outlook voor Android<https://aka.ms/AAb9ysg> downloaden
________________________________

**Reply by divda** [reply]
Beste Hansie,

Uw meter met zijn 78 samples per periode is blijkbaar een andere dan diegene die Peter beschrijft. 
Wat is bij you de telegramgrootte en de bitrate ? Als die veel meer zou zijn dan 2 Mbit/s zou dat de Rx buffer overflow kunnen verklaren.
De mijne is een Sagemcom Siconia S211, waarvan de telegramstructuur en snelheid overeen zouden moeten komen met de simulator van Peter.

Heb je je arduino code getest met je meter en/of de simulator van Peter ? In de simulator kan je de bitrate verlagen om te zien wat de ontvanger wel nog aankan.

Het is mijn bedoeling om voor het ontwikkelen van de ontvanger te starten met arduino, met esp-idf heb ik geen ervaring. Ik hoop dat ik interrupts en dma kan vermijden, want ik heb gehoord dat arduino esp32 op dat gebied niet goed werkt, en dat esp-idf  beter zou zijn.

Als je me de kans geeft, wil ik wel eens naar je arduino code en de instellingen kijken.

Mvg,  Dirk

**Comment by PerformerH** [comment]
Hallo Dirk,

Bij mij is het een T211. Siconia.

De telegrammen zijn identiek aan deze van Peter.
Ik heb geprobeerd met de meter zelf. Baudrate op 2Mb. ik neem niet alle data binnen. Zoals Peter lees ik een volledige sinus en verwerk daarna de gegevens. Met esp idf lukt me dit perfect. Bij arduino heb ik een 20 tal telegrammen en daar wordt mijn buffer overschreven....

Ik denk dat er te weinig tijd tussen de telegrammen zit om de buffers snel genoeg te ledigen.

Alvast veel succes.

Verder FYI,
Momenteel gebruik ik een em24 meter. Deze geeft 2 metingen per seconde. Deze hapert af en toe, vandaar dat ik deze wens te vervangen door een esp32.

Mvg Hansie

Outlook voor Android<https://aka.ms/AAb9ysg> downloaden
________________________________

**Reply by divda** [reply]
Hansie,

Als je een Siconia T211 hebt , hoe kom je dan aan die 78 samples per Hz ? 

Mvg, Dirk

**Comment by PerformerH** [comment]
Hallo Dirk,
Ik dacht eerst dat ik de berichten verkeerd uitlas. Maar in de sampledata staat wel degelijk dat er 78 samples zijn.

Geen idee hoe dit gestuurd wordt of als dit een instelbare parameter is.

Mvg Hansie

Outlook voor Android<https://aka.ms/AAb9ysg> downloaden
________________________________

**Reply by tribp** [reply]
Beste Hansie,
Zie korte uitleg in README of lange uitleg :-) https://github.com/tribp/DSMR-Fluvius-MQTT-Openhab/blob/master/Doc/e-MUCS_H_Ed_1_3.docx

Aangezien de RS422 snelheid fixed is (2Mbps) en telegram = 45 bytes x 10bits/byte x 0.5us/bit = 225 us per telegram  
Opm: de telegram is ALTIJD fixed 225us maar niet de sample freq

1 telegram = 1 (sample)waarde v/d sinus
Vandaar dat men niet sneller mag of kan sampelen dan 4200 Hz of maw om de 1/4200= 238 us want anders riskeer je sneller dan je bussnelheid van je RS422 te sampelen.
Of anders gesteld: Elke waarde duurt sowieso 225us om op de draad te zetten waardoor je niet sneller mag waardes nemen dan je ze  kwijt kan op de draad
Vandaar ook de "pauzes" op de scope tussen de telegrammen

Hopelijk maakt dit iets wat duidelijker. Lees zeker eens de e-Mucs standaard

PS: zie pag 13 van de standaard waarbij byte 16 "Sampling frequency" wordt gedefinieerd (fixed of niet) en wat de waarde betekent

gr Peter

**Comment by PerformerH** [comment]
Hallo Peter,

Dat klopt idd. Ik denk dat 78 samples ongeveer het maximum is. Alles klopt maar îk kan enkel snel genoeg uitlezen met esp idf. Met arduino mis ik de data na een 20 samples.
Ik heb geprobeerd om continue uit te lezen zonder enige andere actie. Zonder resultaat.

Ook met vaste telegramlengte zoals uw voorbeeld, ...

Alvast bedankt voor de respons,
Mvg Hansie.

Outlook voor Android<https://aka.ms/AAb9ysg> downloaden
________________________________

**Reply by divda** [reply]
Hansie,

Heb je die arduino ontvanger van Peter al geprobeerd ? Bij mij doet die een volledig frame zonder data te verliezen. Dat is te zien als je de frame nummers laat afdrukken, die zijn oplopend, zonder onderbrekingen.

**Comment by PerformerH** [comment]
Hallo Dirk,
Ja, deze heb ik geprobeerd. Probeer je dit met je fluviusmeter. Of met de simulator van Peter?
Hoeveel frames heb je per Hz?
Ik gebruik de wroom versie met 4mb.


Alvast bedankt voor de update.

Mvg, hansie

Outlook voor Android<https://aka.ms/AAb9ysg> downloaden
________________________________

**Reply by divda** [reply]
Hansie,

Ik heb het geprobeerd met de fluviusmeter , en pakketten hexadecimaal weergegeven . Het aantal samples is 34 in hex, en dat is 52 decimaal.

**Comment by PerformerH** [comment]
Ik zal deze nog eens proberen. Dat zou ik echt wel de este oplossing vinden.  Arduino is een stuk eenvoudiger.
Bedankt Dirk.


Outlook voor Android<https://aka.ms/AAb9ysg> downloaden
________________________________
